### ISSUE TITLE
计算FLOPs

### ISSUE THREAD

**Issue by zsbjmy** [PaddlePaddle/PaddleDetection#3479]
在tools/infer.py加入如下函数，得到FLOPs值为-1,请问原因？

`def ZSflops(FLAGS, cfg):
    # build trainer
    trainer = Trainer(cfg, mode='test')

    # load weights
    trainer.load_weights(cfg.weights)
    # get inference images
    images = get_test_images(FLAGS.infer_dir, FLAGS.infer_img)
    trainer.dataset.set_images(images)
    loader = create('TestReader')(trainer.dataset, 0)

    imid2path = trainer.dataset.get_imid2path()

    anno_file = trainer.dataset.get_anno()
    clsid2catid, catid2name = get_categories(trainer.cfg.metric, anno_file)
    # Run Infer 
    trainer.status['mode'] = 'test'
    trainer.model.eval()
    for step_id, data in enumerate(loader):
        trainer.status['step_id'] = step_id
        # print(data)
        # forward
        flops0 = paddle.flops(trainer.model(data), [1, 3, 608, 608],print_detail=True)
        print(flops0)
    return flops0`

**Comment by wangxinxin08** [comment]
目前这个API可能存在问题

**Comment by yghstill** [comment]
@zsbjmy 已提交PR支持FLOPs计算：https://github.com/PaddlePaddle/PaddleDetection/pull/3629

**Comment by lzk901372** [comment]
现在我暂时弄了一个可以打印模型FLOPs和Params的办法。思路和解决方式依然是使用Paddle的原生的计算FLOPs的方法，但是不敢保证计算的结果一定是最正确的，欢迎讨论。

首先Paddle的FLOPs计算代码和Pytorch里的thop提供的计算方式如出一辙，仔细看一下源代码就可以看出来，可能最明显的区别就在于，PaddleDetection是一个封装得很好的框架，一个模型（比如trainer.model）的输入不是一个Tensor而是一个dict对象，其中键image对应的值才是输入图像数据的Tensor，而在thop的profile当中一般一个模型的输入就是一个Tensor（其实PaddleDetection模型在forward的时候也是用键image的值，不是用整个dict……）。因此说到底，PaddleDetection模型在向前传播的时候需要一个封装好的程序字典对象作为input，这就和Paddle提供的flops函数的形参inputs要求的数据类型不对应。

但是当我进入flops函数的内部去看的时候，发现他调用了一个更深层的函数dynamic_flops（这里我们先不讨论静态图的FLOPs计算）来计算FLOPs，而flops这个函数其实是在判断形参inputs是不是nn.Layer类型，如下所示：
```
def flops(net, input_size, custom_ops=None, print_detail=False):
     if isinstance(net, nn.Layer):
            inputs = paddle.randn(input_size)
            return dynamic_flops(
                net,
                inputs=inputs,
                custom_ops=custom_ops,
                print_detail=print_detail)
        elif isinstance(net, paddle.static.Program):
            return static_flops(net, print_detail=print_detail)
        else:
            warnings.warn(
                "Your model must be an instance of paddle.nn.Layer or paddle.static.Program."
            )
            return -1
```
说明flops这个函数表面只是用来判断形参类型的，真正在干事情的是dynamic_flops这个函数。这一点和thop里的profile函数不一样。所以我们只需要想办法修改flops函数或者直接调用dynamic_flops函数就可以计算模型的FLOPs和Params，并且过程和thop的profile函数流程基本一致。

我的做法可能比较笨，我直接在 `ppdet/engine/` 文件夹下新建了一个.py文件，把[https://github.com/PaddlePaddle/Paddle/blob/e04b66f2d272d68f77dcd94cb2956938475411d8/python/paddle/hapi/dynamic_flops.py#L205](url)这个文件里除掉import和flops函数的其他内容基本都复制了过去，然后把dynamic_flops函数改成了一个我自己想要的名字（cal_flops）。计算FLOPs我是这样用的：在运行infer.py时会转到trainer.py执行predict函数（执行infer能运行test模型，保证batch为1），在enumerate数据集我加了一条：
```
......
        for step_id, data in enumerate(loader):
            self.status['step_id'] = step_id

            # Calculation of Params and FLOPs using the first input data
            if step_id == 0:
                flops, params = MF(self.model, data)  # <-MF是我封装的一个计算FLOPs和Params的函数，但是只需要像profile函数一样，传入模型和输入即可。这里的data是一个dict对象

            # forward
            outs = self.model(data)

            for key in ['im_shape', 'scale_factor', 'im_id']:
                outs[key] = data[key]
            for key, value in outs.items():
                outs[key] = value.numpy()
......
```
反正，只要知道了函数该怎么改，计算方法又是相似的，FLOPs是可以计算出来的。我计算了一遍SSD300-VGG16的GFLOPS，算的是30.35674GFLOPs，不知道对不对，不过跟网上搜到的一些结果对比而言好像差距不大。

顺便提一句，Params的计算应该比较简单，直接用numel函数嵌套递归就可以实现。

**Comment by zsbjmy** [comment]
非常感谢~发自我的荣耀手机-------- 原始邮件 --------发件人： Kaleidoscope ***@***.***>日期： 2021年8月26日周四 21:03收件人： PaddlePaddle/PaddleDetection ***@***.***>抄送： zsbjmy ***@***.***>, Mention ***@***.***>主    题： Re: [PaddlePaddle/PaddleDetection] 计算FLOPs (#3479)
现在我暂时弄了一个可以打印模型FLOPs和Params的办法。思路和解决方式依然是使用Paddle的原生的计算FLOPs的方法，但是不敢保证计算的结果一定是最正确的，欢迎讨论。
首先Paddle的FLOPs计算代码和Pytorch里的thop提供的计算方式如出一辙，仔细看一下源代码就可以看出来，可能最明显的区别就在于，PaddleDetection是一个封装得很好的框架，一个模型（比如trainer.model）的输入不是一个Tensor而是一个dict对象，其中键image对应的值才是输入图像数据的Tensor，而在thop的profile当中一般一个模型的输入就是一个Tensor（其实PaddleDetection模型在forward的时候也是用键image的值，不是用整个dict……）。因此说到底，PaddleDetection模型在向前传播的时候需要一个封装好的程序字典对象作为input，这就和Paddle提供的flops函数的形参inputs要求的数据类型不对应。
但是当我进入flops函数的内部去看的时候，发现他调用了一个更深层的函数dynamic_flops（这里我们先不讨论静态图的FLOPs计算）来计算flops，而flops这个函数其实是在判断形参inputs是不是nn.Layer类型，如下所示：
def flops(net, input_size, custom_ops=None, print_detail=False):
     if isinstance(net, nn.Layer):
            inputs = paddle.randn(input_size)
            return dynamic_flops(
                net,
                inputs=inputs,
                custom_ops=custom_ops,
                print_detail=print_detail)
        elif isinstance(net, paddle.static.Program):
            return static_flops(net, print_detail=print_detail)
        else:
            warnings.warn(
                "Your model must be an instance of paddle.nn.Layer or paddle.static.Program."
            )
            return -1

说明flops这个函数表面只是用来判断形参类型的，真正在干事情的是dynamic_flops这个函数。这一点喝thop里的profile函数不一样。所以我们只需要想办法修改flops函数或者直接调用dynamic_flops函数就可以计算模型的FLOPs和Params，并且过程和thop的profile函数流程基本一致。
我的做法可能比较笨，我直接在 ppdet/engine/ 文件夹下新建了一个.py文件，把https://github.com/PaddlePaddle/Paddle/blob/e04b66f2d272d68f77dcd94cb2956938475411d8/python/paddle/hapi/dynamic_flops.py#L205这个文件里除掉import和flops函数的其他内容基本都复制了过去，然后把dynamic_flops函数改成了一个我自己想要的名字（cal_flops）。计算FLOPs我是这样用的：在运行infer.py时会转到trainer.py执行predict函数（执行infer能运行test模型，保证batch为1），在enumerate数据集我加了一条：
......
        for step_id, data in enumerate(loader):
            self.status['step_id'] = step_id

            # Calculation of Params and FLOPs using the first input data
            if step_id == 0:
                flops, params = MF(self.model, data)  # <-MF是我封装的一个计算FLOPs和Params的函数，但是只需要像profile函数一样，传入模型和输入即可。这里的data是一个dict对象

            # forward
            outs = self.model(data)

            for key in ['im_shape', 'scale_factor', 'im_id']:
                outs[key] = data[key]
            for key, value in outs.items():
                outs[key] = value.numpy()
......

反正，只要知道了函数该怎么改，计算方法又是相似的，FLOPs是可以计算出来的。我计算了一遍SSD300-VGG16的GFLOPS，算的是30.35674GFLOPs，不知道对不对，不过跟网上搜到的一些结果对比而言好像差距不大。
顺便提一句，Params的计算应该比较简单，直接用numel函数嵌套递归就可以实现。

**Comment by lzk901372** [comment]
顺便提一下，Paddle里原生添加计算FLOPs和Params的hooker的数据类型有问题。原始hooker初始化代码是这样的：
```
 def add_hooks(m):
        if len(list(m.children())) > 0:
            return
        m.register_buffer('total_ops', paddle.zeros([1], dtype='int32'))
        m.register_buffer('total_params', paddle.zeros([1], dtype='int32'))
        m_type = type(m)
        ......
```
但是实际buffer数据大小很可能超过int32的上限。我在计算上面SSD的FLOPs时就遇到过某一个sublayer的total_ops变成负数的（最终算出来的FLOPs只有1GFLOPs左右，很显然不正确），然而把dtype改成int64或者float64就没有问题了。我建议对照thop里的做法，改成float64。
